File: source/rendering/lodControl.cpp
/*
 * This file is part of Cosmonium.
 *
 * Copyright (C) 2018-2021 Laurent Deru.
 *
 * Cosmonium is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Cosmonium is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Cosmonium.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "quadTreeNode.h"
#include "lodControl.h"


LodControl::~LodControl(void)
{
}

void
LodControl::set_texture_size(unsigned int texture_size)
{
}

bool
LodControl::should_instanciate(QuadTreeNode *patch, double apparent_patch_size, double distance)
{
  return patch->visible && patch->children.size() == 0;
}

bool
LodControl::should_remove(QuadTreeNode *patch, double apparent_patch_size, double distance)
{
  return !patch->visible;
}

TextureLodControl::TextureLodControl(unsigned int min_density, unsigned int density, unsigned int max_lod) :
    min_density(min_density),
    density(density),
    max_lod(max_lod),
    texture_size(0)
{
}

void
TextureLodControl::set_texture_size(unsigned int texture_size)
{
  this->texture_size = texture_size;
}

unsigned int
TextureLodControl::get_density_for(unsigned int lod)
{
  return max(min_density, density / (1 << lod));
}

bool
TextureLodControl::should_split(QuadTreeNode *patch, double apparent_patch_size, double distance)
{
  if (patch->lod < max_lod) {
    return (texture_size > 0 && apparent_patch_size > texture_size * 1.1); //TODO: add appearance.texture.can_split(patch)
  } else {
    return false;
  }

}

bool
TextureLodControl::should_merge(QuadTreeNode *patch, double apparent_patch_size, double distance)
{
  return (apparent_patch_size < texture_size / 1.1);
}

TextureOrVertexSizeLodControl::TextureOrVertexSizeLodControl(unsigned int max_vertex_size, unsigned int min_density, unsigned int density, unsigned int max_lod) :
    TextureLodControl(min_density, density, max_lod),
    max_vertex_size(max_vertex_size)
{
}

bool
TextureOrVertexSizeLodControl::should_split(QuadTreeNode *patch, double apparent_patch_size, double distance)
{
  if (patch->lod < max_lod) {
    if (texture_size > 0) {
        return apparent_patch_size > texture_size * 1.1;
    } else {
        double apparent_vertex_size = apparent_patch_size / patch->density;
        return apparent_vertex_size > max_vertex_size;
    }
  } else {
    return false;
  }
}

bool
TextureOrVertexSizeLodControl::should_merge(QuadTreeNode *patch, double apparent_patch_size, double distance)
{
  if (texture_size > 0) {
      return apparent_patch_size < texture_size / 1.1;
  } else {
      double apparent_vertex_size = apparent_patch_size / patch->density;
      return apparent_vertex_size < max_vertex_size / 1.1;
  }
}

VertexSizeLodControl::VertexSizeLodControl(unsigned int max_vertex_size, unsigned int density, unsigned int max_lod) :
    density(density),
    max_lod(max_lod),
    max_vertex_size(max_vertex_size)
{
}

unsigned int
VertexSizeLodControl::get_density_for(unsigned int lod)
{
  return density;
}

bool
VertexSizeLodControl::should_split(QuadTreeNode *patch, double apparent_patch_size, double distance)
{
  if (patch->lod < max_lod) {
      double apparent_vertex_size = apparent_patch_size / patch->density;
      return apparent_vertex_size > max_vertex_size * 1.1;
  } else {
      return false;
  }
}

bool
VertexSizeLodControl::should_merge(QuadTreeNode *patch, double apparent_patch_size, double distance)
{
    double apparent_vertex_size = apparent_patch_size / patch->density;
    return apparent_vertex_size < max_vertex_size / 1.1;
}

VertexSizeMaxDistanceLodControl::VertexSizeMaxDistanceLodControl(double max_distance, unsigned int max_vertex_size, unsigned int density, unsigned int max_lod) :
    VertexSizeLodControl(max_vertex_size, density, max_lod),
    max_distance(max_distance)
{
}

bool
VertexSizeMaxDistanceLodControl::should_instanciate(QuadTreeNode *patch, double apparent_patch_size, double distance)
{
  return patch->visible && distance < max_distance;
}

bool
VertexSizeMaxDistanceLodControl::should_remove(QuadTreeNode *patch, double apparent_patch_size, double distance)
{
  return !patch->visible;
}

File: cosmonium/pyrendering/lodcontrol.py
#
# This file is part of Cosmonium.
#
# Copyright (C) 2018-2024 Laurent Deru.
#
# Cosmonium is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Cosmonium is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Cosmonium.  If not, see <https://www.gnu.org/licenses/>.
#


class LodControl(object):
    def __init__(self, density=32, max_lod=100):
        self.density = density
        self.max_lod = max_lod

    def get_density_for(self, lod):
        return self.density

    def set_appearance(self, appearance):
        pass

    def set_texture_size(self, appearance):
        pass

    def should_split(self, patch, apparent_patch_size, distance):
        return False

    def should_merge(self, patch, apparent_patch_size, distance):
        return False

    def should_instanciate(self, patch, apparent_patch_size, distance):
        return patch.visible and len(patch.children) == 0

    def should_remove(self, patch, apparent_patch_size, distance):
        return not patch.visible


# The lod control classes uses hysteresis to avoid cycle of split/merge due to
# precision errors.
# When splitting the resulting patch will be 1.1 bigger than the merge limit
# When merging, the resulting patch will be  1.1 smaller than the slit limit


class TextureLodControl(LodControl):
    def __init__(self, min_density, density, max_lod=100):
        LodControl.__init__(self, density, max_lod)
        self.min_density = min_density
        self.texture_size = 0

    def get_density_for(self, lod):
        return max(self.min_density, self.density // (1 << lod))

    def set_appearance(self, appearance):
        self.appearance = appearance
        if appearance is not None and appearance.texture is not None:
            self.texture_size = appearance.texture.source.texture_size
        else:
            self.texture_size = 0

    def set_texture_size(self, texture_size):
        self.texture_size = texture_size

    def should_split(self, patch, apparent_patch_size, distance):
        if patch.lod < self.max_lod:
            return (
                self.texture_size > 0 and apparent_patch_size > self.texture_size * 1.1
            )  # and self.appearance.texture.can_split(patch)
        else:
            return False

    def should_merge(self, patch, apparent_patch_size, distance):
        return apparent_patch_size < self.texture_size / 1.1


class TextureOrVertexSizeLodControl(TextureLodControl):
    def __init__(self, max_vertex_size, min_density, density, max_lod=100):
        TextureLodControl.__init__(self, min_density, density, max_lod)
        self.max_vertex_size = max_vertex_size

    def should_split(self, patch, apparent_patch_size, distance):
        if patch.lod >= self.max_lod:
            return False
        if self.texture_size > 0:
            if apparent_patch_size > self.texture_size * 1.1:
                return True
        else:
            apparent_vertex_size = apparent_patch_size / patch.density
            return apparent_vertex_size > self.max_vertex_size

    def should_merge(self, patch, apparent_patch_size, distance):
        if self.texture_size > 0:
            return apparent_patch_size < self.texture_size / 1.1
        else:
            apparent_vertex_size = apparent_patch_size / patch.density
            return apparent_vertex_size < self.max_vertex_size / 1.1


class VertexSizeLodControl(LodControl):
    def __init__(self, max_vertex_size, density, max_lod=100):
        LodControl.__init__(self, density, max_lod)
        self.max_vertex_size = max_vertex_size

    def should_split(self, patch, apparent_patch_size, distance):
        if patch.lod < self.max_lod:
            apparent_vertex_size = apparent_patch_size / patch.density
            return apparent_vertex_size > self.max_vertex_size * 1.1
        else:
            return False

    def should_merge(self, patch, apparent_patch_size, distance):
        apparent_vertex_size = apparent_patch_size / patch.density
        return apparent_vertex_size < self.max_vertex_size / 1.1


class VertexSizeMaxDistanceLodControl(VertexSizeLodControl):
    def __init__(self, max_distance, max_vertex_size, density, max_lod=100):
        VertexSizeLodControl.__init__(self, max_vertex_size, density, max_lod)
        self.max_distance = max_distance

    def should_instanciate(self, patch, apparent_patch_size, distance):
        return patch.visible and distance < self.max_distance

    def should_remove(self, patch, apparent_patch_size, distance):
        return not patch.visible
